[FEAT] Instagram Hashtag 검색 결과 크롤링

DunDung opened the pull request:
- Hastag 개수
- 인기 게시물 9개 정보
    - 이미지 URL
    - 포스트 URL 
- 프록시 설정

fixes: #17

DunDung committed:
feat: proxy 추가

ordinCode committed:
feat : InstagramCrawler.java

 - 인기게시물 부분 발취

DunDung committed:
refactor: hash-tag count 리팩토링

DunDung committed:
style: 안쓰는 상수 삭제

ordinCode committed:
refactor : CrawlingProcessor.java

 - 정제하는 메서드 분리

DunDung committed:
refactor: Document 추출 로직 수정 및 테스트 추가

 - 인덱스로 찾던 것을 정규식으로 수정
 - 코드 중복 제거
 - 테스트 추가

DunDung committed:
refactor: 네이밍 변경 및 메소드 분리, 코드 정리

ordinCode committed:
feat: CrawlingDto 생성 구현

 - 전달받은 가게 이름 기준으로 CrawlingDto 생성

DunDung committed:
refactor: 메소드 분리 및 Crawler timeout 추가

ordinCode committed:
feat: InstagramCrawlerTest 생성

DunDung committed:
test: ProxyFactory, Proxy 테스트 추가

ordinCode committed:
feat: InstagramCrawlerTest 수정

 - assertAll적용

DunDung committed:
refactor: List<PostDto> 일급 컬렉션으로 변경 및 테스트

DunDung committed:
style: 메소드, 변수 네이밍 수정

ordinCode committed:
refactor: Extractor.extractByKey 메서드 추가

 - InstagramCrawler 의 makeDtos에서 Extractor.extractByKey를 사용하도록 리펙토링

DunDung committed:
refactor: CustomException으로 변경 및 테스트

ordinCode committed:
refactor: Pattern을 emun으로 전환

DunDung committed:
refactor: RegexPattern Test 추가

ordinCode reviewed (approved):
페어프로그래밍으로 인한 approve

ordinCode committed:
feat: README.md 추가

ordinCode committed:
feat: README.md 추가

giantim commented on hashtagmap-instagram-crawler/build.gradle:
```diff
@@ -2,4 +2,6 @@ bootJar { enabled = false }
 
 dependencies {
+    implementation 'org.jsoup:jsoup:1.13.1'
+    compile group: 'com.google.code.gson', name: 'gson', version: '2.8.6'
```
compile을 implementation으로 바꾸는 것은 어떨까요 :)
```groovy
implementation 'com.google.code.gson:gson:2.8.6'
```

ordinCode replied:
앗 실수~ 꼼꼼한 지적 👍

giantim commented on hashtagmap-instagram-crawler/src/main/java/com/songpapeople/hashtagmap/crawler/InstagramCrawler.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class InstagramCrawler {
```
불필요한 공백 한 줄이 있네요 :)

giantim reviewed (commented):
둔덩 학성! 크롤러 구현 하느라 고생 많으셨습니다~~!
핵심 모듈인데 기능도 짧고 정확하게 잘 구현하시고 테스트 케이스도 다양하게 추가되어있어서 이해하기 편했네요 :)
간단한 의견 남겼습니다!
수고하셨습니다 :)

ordinCode committed:
refactor: compile을 implementation로 변경

moonyoungCHAE commented on hashtagmap-instagram-crawler/src/test/java/com/songpapeople/InstagramCrawlerApplicationTest.java:
```diff
@@ -0,0 +1,11 @@
+@SpringBootTest
+public class InstagramCrawlerApplicationTest {
+
+    public void contextLoads() {
```
요 메서드는 없어도 될 것 같습니다!

ordinCode replied:
그러네요 다시 확인해보니 Spring환경에서 하는 테스트가 없네요.

moonyoungCHAE commented on hashtagmap-instagram-crawler/src/main/java/com/songpapeople/hashtagmap/exception/CrawlingUrlException.java:
```diff
@@ -0,0 +1,8 @@
+public class CrawlingUrlException extends RuntimeException {
+
+    public CrawlingUrlException() {
+        super("연결할 수 없는 url 입니다.");
```
RuntimeException 를 throw 하게 되는데 보다 구체적인  Exception을 던져주면 어떨까요??
extends IllegalArgumentException 를 제안합니다 ;)

ordinCode replied:
이유는 무엇인가요? 저는 딱히 IllegalArgumentException로 익셉션의 의미를 한정시키고 싶지 않아서 RuntimeException으로 했는데 혹시 기술적인 이슈가 있을까요?

moonyoungCHAE replied:
기술적인 이슈는 없어요!
다만 좀 더 명확한 Exception을 주는 게 어떤가 생각했어요 :)
통합적인 Custom Exception이라면 Runtime Exception을,
구체적인 Custom Exception이라면 구체적인 Exception을 주는 게 어떨까 생각이 드네요..

팀 내의 Custom Exception 규칙을 정할 때 요 이슈도 정하면 되겠네요 ^___^

moonyoungCHAE commented on hashtagmap-instagram-crawler/src/main/java/com/songpapeople/hashtagmap/exception/NotFoundRegexException.java:
```diff
@@ -0,0 +1,8 @@
+public class NotFoundRegexException extends RuntimeException {
+
+    public NotFoundRegexException() {
+        super("원하는 내용을 찾지 못했습니다.");
```
여기도 마찬가지로 IllegalArgumentException  제안합니다~

moonyoungCHAE commented on hashtagmap-instagram-crawler/src/test/java/com/songpapeople/hashtagmap/dto/PostDtosTest.java:
```diff
@@ -0,0 +1,22 @@
+    @Test
+    void createOverSize() {
+        List<PostDto> postDtos = new ArrayList<>();
+        for (int i = 1; i <= 10; i++) {
```
경계값 테스트 👍 
1. 다만 10이라는 숫자에 변수로써 의미를 부여한다면 더 직관적인 테스트가 되지 않을까 라는 생각이 들었어요!
PostDtos.MAX_POPULAR_POST_SIZE + 1; 와 같이

2. 8개일 때도 테스트를 진행한다면 좋을 것 같아요

moonyoungCHAE commented on hashtagmap-instagram-crawler/src/main/java/com/songpapeople/hashtagmap/crawler/Crawler.java:
```diff
@@ -0,0 +1,25 @@
+        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36";
+    private static final int HOLDING_TIME = 7000;
+
+    public static Document crawling(String url) {
```
자주 쓰이는 요 메서드를 static으로 뺀 것 💯  좋네욤~

moonyoungCHAE commented on hashtagmap-instagram-crawler/src/test/java/com/songpapeople/hashtagmap/proxy/ProxyTest.java:
```diff
@@ -0,0 +1,25 @@
+        String actualHost = System.getProperty("http.proxyHost");
+        String actualPort = System.getProperty("http.proxyPort");
+
+        assertAll(
```
assertAll 이란 것도 있군요~ 한 수 배워갑니다.

moonyoungCHAE reviewed (approved):
크롤링은 처음 접하는 분야인데도 빠르게 적응해서 잘 구현하셨네요.
코드도 깔끔하니 잘 읽히고, 테스트도 풍부하네요!!! 💯 
몇 가지 코멘트 남겼으니 확인 부탁드려요 :)

ordinCode committed:
refactor: Spring contextloads 메서드 제거

pci2676 commented on hashtagmap-instagram-crawler/src/main/java/com/songpapeople/hashtagmap/crawler/Crawler.java:
```diff
@@ -0,0 +1,25 @@
+
+import com.songpapeople.hashtagmap.exception.CrawlingUrlException;
+
+public class Crawler {
```
타 모듈로 노출되지 않아도 되는 클래스라면 protected 혹은 package-private 접근제어자로 바꿔보는건 어떨까요?
이 모듈을 사용하는 측에서는 InstagramCrawler만 알고 다른 클래스는 감춰져 있는게 나을 수 있다는 생각이 문득 드네요!

pci2676 commented on hashtagmap-instagram-crawler/src/test/java/com/songpapeople/hashtagmap/crawler/JsonExplorerTest.java:
```diff
@@ -0,0 +1,17 @@
+
+class JsonExplorerTest {
+    @Test
+    void findByKey() {
```
물론 테스트 코드의 메서드 이름으로 테스트 코드의 목적을 명시해 주는 방법도 좋지만
개인적으로 `@DisplayName` 을 이용해서 한글로 테스트 코드의 목적을 적어주는 방법을 더 선호합니다!
아무래도 한글을 사용하면 조금 더 명시적으로 테스트의 목적을 적어 줄 수 있다고 생각이 들거든요!

pci2676 replied:
[자바와 JUnit을 활용한 실용주의 단위 테스트](http://www.yes24.com/Product/Goods/75189146) 
**4.5 문서로서의 테스트**의 
**4.5.1 일관성 있는 이름으로 테스트 문서화** 부분을 보면 다음과 같은 이야기가 있습니다.

> 테스트하려는 맥락을 제안하기 보다는 어떤 맥락에서 일련의 행동을 호출했을 때 어떤 결과가 나오는지를 명시하세요.

그리고 영어로 테스트 코드 메서드의 명명 예시를 보여주는데 다음과 같습니다.

doingSomeOperationGeneratesSomeResult : 어떤 동작을 하면 어떤 결과가 나온다.
someResultOccursUnderSomeCondition : 어떤 결과는 어떤 조건에 발생한다.

위와 같이 하기가 힘들다면 BDD(행동 주도 개발)에서 말하는 given-when-then 템플릿을 이용해서 테스트의 의도를 조금 더 명확하게 드러낸다면 테스트 코드가 문서로써 역할을 잘 수행 할 수 있지 않을까요?

pci2676 commented on hashtagmap-instagram-crawler/src/main/java/com/songpapeople/hashtagmap/crawler/InstagramCrawler.java:
```diff
@@ -0,0 +1,40 @@
+        Document document = Crawler.crawling(String.format(INSTAGRAM_URL_FORMAT, placeName));
+        String body = document.body().toString();
+        String hashtagCount = RegexPattern.HASH_TAG_COUNT.extract(body);
+        String popularPostsInfo = RegexPattern.HASHTAG_POPULAR_POSTS_INFO.extract(body);
```
`Crawler`로부터 `Document`를 만드는 방식보다 **특정 객체**를 생성해서
해당 객체에서 `getHashtagCount(RegexPattern regexPattern)`, `getPopularPostInfo(RegexPattern regexPattern)` 하는 방식으로 하면 좀 더 테스트하기 용이한 코드가 작성 될 수 있지 않을까요?
지금 코드는 내부 구현의 노출이 많이 되어 있는 것 같아요! 
조금 더 추상화하면 좋을 것 같습니다!

pci2676 replied:
```suggestion
        Foo foo = Crawler.crawling(String.format(INSTAGRAM_URL_FORMAT, placeName));
        String hashtagCount = foo.getHashtagCount();
        String popularPostsInfo = foo.getPopularPostInfo();
```

위와 같은 모습이 될 수 있을 것 같네요!

pci2676 replied:
아니면 더 나아가 `getPopularPostsInfo()` 가 `postDto` 를 반환해도 좋을 것 같다고 생각이 드네요!
제 생각에는 `postDto`는 DTO라고 보기 보다 도메인 영역에 해당하는 객체라는 생각이 듭니다!
`PopularPost` 라는 도메인 객체라고 보는건 어떻게 생각하시나요? 궁금하네요!

pci2676 commented on hashtagmap-instagram-crawler/src/test/java/com/songpapeople/hashtagmap/crawler/RegexPatternTest.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Test
+    void extractHashtagCount() {
+        String extracted = RegexPattern.HASH_TAG_COUNT.extract("kimchi\"edge_hashtag_to_media\":{\"count\":7777, asd}");
```
JSON 형식으로된 테스트용 데이터를 만들기 불편하지 않았나요??
제 경우에는 JSON 포맷을 맞추기 굉장히 귀찮고 힘들었던거로 기억이 나는데요!
`.json`파일을 만들고 테스트 데이터를 관리하고 해당 파일을 읽어서 String 으로 변환 후 테스트하는 방법도 고려해 보면 좋을 것 같아요!

pci2676 reviewed (changes requested):
학성, 둔덩 코드 작성하느라 고생 많으셨습니다!
다같이 고민해볼만한 부분에 대해 리뷰를 남겼습니다!
학성과 둔덩의 답변을 듣고 싶은 부분이 있어서 바로 Approve는 하지 않겠습니다 ㅎㅎ

+추가적으로 나중에 네트워크를 타지 않고 테스트가 가능하도록 프로덕션 코드가 개선되면 좋겠네요!

DunDung committed:
refactor: 접근 제어자 변경 및 PostDtos 테스트 추가

ordinCode committed:
refactor: Test에 DisplayName추가

ordinCode committed:
refactor: Crawling 동작 방식 변경

 - Crawler 결과가 body를 리턴하도록
 - InstagramCrawler.createHashtagDto 의 로직을 InstaCrawlingResult로 분리

ordinCode committed:
feat: InstaCrawlingResult 테스트 추가

ordinCode committed:
refactor: 크롤링 관련 Test Disabled 처리
